src/PelCanonMakerNotes.php

```
@@ -0,0 +1,308 @@
+            $type = $this->data->getShort($this->offset + 12 * $i + 2);
+            $components = $this->data->getLong($this->offset + 12 * $i + 4);
+            $data = $this->data->getLong($this->offset + 12 * $i + 8);
+            // check if tag is defined
+            if (in_array($tag, $this->undefinedMakerNotesTags)) {
+                continue;
+            }
+            switch ($tag) {
+                case PelTag::CANON_CAMERA_SETTINGS:
+                    $this->parseCameraSettings($mkNotesIfd, $this->data, $data, $components);
+                    break;
+                case PelTag::CANON_SHOT_INFO:
+                    $this->parseShotInfo($mkNotesIfd, $this->data, $data, $components);
+                    break;
+                case PelTag::CANON_PANORAMA:
+                    $this->parsePanorama($mkNotesIfd, $this->data, $data, $components);
+                    break;
+                case PelTag::CANON_PICTURE_INFO:
+                    $this->parsePictureInfo($mkNotesIfd, $this->data, $data, $components);
+                    break;
+                case PelTag::CANON_FILE_INFO:
+                    $this->parseFileInfo($mkNotesIfd, $this->data, $data, $components);
+                    break;
+                case PelTag::CANON_CUSTOM_FUNCTIONS:
+                    //TODO
+                default:
+                    $mkNotesIfd->loadSingleValue($this->data, $this->offset, $i, $tag);
+                    break;
+            }
+        }
+        $this->parent->addSubIfd($mkNotesIfd);
+    }
+
+    private function parseCameraSettings($parent, $data, $offset, $components)
+    {
+        $type = PelIfd::CANON_CAMERA_SETTINGS;
+        Pel::debug('Found Canon Camera Settings sub IFD at offset %d', $offset);
+        $size = $data->getShort($offset);
+        $offset += 2;
+        $elemSize = PelFormat::getSize(PelFormat::SSHORT);
+        if ($size / $components !== $elemSize) {
+            // TODO throw
```

Should an exception be thrown in this case (and in similar other cases?)

---

You're right. Embarrassing I missed the TODOs...
One problem is, that I don't know if it is correct to throw an exception at this point. Since there are unknown tags which could be of another format. Thus it would throw an `PelMalformedMakerNotesException` or something else, just because there is an unknown tag.
What's your opinion? I'm ok with both solutions (1. remove the checks and 2. keep them and adjust if there is a case in the future where the exception is thrown although it is a valid maker notes section).

---

I would throw an exception so we would get bug reports if there is something wrong. Code could be improved. Otherwise strange results could be seen. I haven't worked with those parsers yet, so I don't really knw about the impact.

---

Ok, will add a exception the next days. Do you refer to a specific part for improvement?

---

My idea is that we might see exception reports when users report PelMalformedMakerNotesException .

---

Please have a look at the new scruntizer tests which show 43 new issues and a decrease in code coverage.

---

Most of the issues are not caused by this PR. Code Coverage is difficult. E.g. `PelEntrySShort` got from 100% code coverage to ~2%, because I had to add the `getText` method. It should be possible to increase it a bit, but I added more than 1000 Lines and almost 700 are for different lenses. Thus I'm not sure how much it could be increased and what would be a good goal.

And since scrutinizer was added today I'm not sure if the result are representative in any way. But I'll have a look at the issues/code coverage caused by this PR.